File: homework-1/nodejs/src/validators.js
// Validation utilities for transactions and accounts
const VALID_CURRENCIES = [
  'USD', 'EUR', 'GBP', 'JPY', 'AUD', 'CAD', 'CHF', 'CNY', 'INR', 'MXN',
  'SGD', 'HKD', 'NZD', 'SEK', 'NOK', 'DKK', 'ZAR', 'BRL', 'RUB', 'KRW'
];

const ACCOUNT_FORMAT = /^ACC-[A-Z0-9]{5}$/;
const TRANSACTION_TYPES = ['deposit', 'withdrawal', 'transfer'];

/**
 * Validates transaction data
 * @param {Object} transaction - The transaction object to validate
 * @returns {Object} - { isValid: boolean, errors: Array }
 */
function validateTransaction(transaction) {
  const errors = [];

  // Validate fromAccount
  if (!transaction.fromAccount) {
    errors.push({ field: 'fromAccount', message: 'fromAccount is required' });
  } else if (!ACCOUNT_FORMAT.test(transaction.fromAccount)) {
    errors.push({ field: 'fromAccount', message: 'Account number must follow format ACC-XXXXX (where X is alphanumeric)' });
  }

  // Validate toAccount
  if (!transaction.toAccount) {
    errors.push({ field: 'toAccount', message: 'toAccount is required' });
  } else if (!ACCOUNT_FORMAT.test(transaction.toAccount)) {
    errors.push({ field: 'toAccount', message: 'Account number must follow format ACC-XXXXX (where X is alphanumeric)' });
  }

  // Validate amount
  if (transaction.amount === undefined || transaction.amount === null) {
    errors.push({ field: 'amount', message: 'amount is required' });
  } else if (typeof transaction.amount !== 'number' || transaction.amount <= 0) {
    errors.push({ field: 'amount', message: 'Amount must be a positive number' });
  } else if (!/^\d+(\.\d{1,2})?$/.test(transaction.amount.toString())) {
    errors.push({ field: 'amount', message: 'Amount must have maximum 2 decimal places' });
  }

  // Validate currency
  if (!transaction.currency) {
    errors.push({ field: 'currency', message: 'currency is required' });
  } else if (!VALID_CURRENCIES.includes(transaction.currency.toUpperCase())) {
    errors.push({ field: 'currency', message: 'Invalid currency code. Use valid ISO 4217 codes (e.g., USD, EUR, GBP)' });
  }

  // Validate type
  if (!transaction.type) {
    errors.push({ field: 'type', message: 'type is required' });
  } else if (!TRANSACTION_TYPES.includes(transaction.type)) {
    errors.push({ field: 'type', message: 'Type must be one of: deposit, withdrawal, transfer' });
  }

  // Validate that fromAccount and toAccount are different for transfers
  if (transaction.type === 'transfer' && transaction.fromAccount === transaction.toAccount) {
    errors.push({ field: 'accounts', message: 'fromAccount and toAccount cannot be the same for transfers' });
  }

  return {
    isValid: errors.length === 0,
    errors
  };
}

module.exports = {
  validateTransaction,
  VALID_CURRENCIES,
  ACCOUNT_FORMAT,
  TRANSACTION_TYPES
};

File: homework-1/python/src/validators.py
"""Validation utilities for Banking Transactions API."""

import re
from typing import Optional
from .models import TransactionCreate, TransactionType, VALID_CURRENCIES, ValidationErrorDetail


# Account format: ACC-XXXXX where X is alphanumeric (uppercase)
ACCOUNT_PATTERN = re.compile(r'^ACC-[A-Z0-9]{5}$')


def validate_account_format(account: Optional[str], field_name: str) -> Optional[ValidationErrorDetail]:
    """Validate account number format."""
    if account is None:
        return None
    if not ACCOUNT_PATTERN.match(account):
        return ValidationErrorDetail(
            field=field_name,
            message=f"Account must follow format ACC-XXXXX (5 alphanumeric characters). Got: {account}"
        )
    return None


def validate_amount(amount: float) -> Optional[ValidationErrorDetail]:
    """Validate transaction amount."""
    if amount <= 0:
        return ValidationErrorDetail(
            field="amount",
            message="Amount must be a positive number"
        )

    # Check for maximum 2 decimal places
    decimal_str = str(amount)
    if '.' in decimal_str:
        decimal_places = len(decimal_str.split('.')[1])
        if decimal_places > 2:
            return ValidationErrorDetail(
                field="amount",
                message="Amount must have maximum 2 decimal places"
            )
    return None


def validate_currency(currency: str) -> Optional[ValidationErrorDetail]:
    """Validate ISO 4217 currency code."""
    if currency.upper() not in VALID_CURRENCIES:
        return ValidationErrorDetail(
            field="currency",
            message=f"Invalid currency code. Valid codes: {', '.join(sorted(VALID_CURRENCIES))}"
        )
    return None


def validate_transaction(transaction: TransactionCreate) -> list[ValidationErrorDetail]:
    """
    Validate a transaction request.

    Returns a list of validation errors (empty if valid).
    """
    errors: list[ValidationErrorDetail] = []

    # Validate amount
    amount_error = validate_amount(transaction.amount)
    if amount_error:
        errors.append(amount_error)

    # Validate currency
    currency_error = validate_currency(transaction.currency)
    if currency_error:
        errors.append(currency_error)

    # Validate based on transaction type
    if transaction.type == TransactionType.DEPOSIT:
        # Deposit requires toAccount
        if not transaction.toAccount:
            errors.append(ValidationErrorDetail(
                field="toAccount",
                message="toAccount is required for deposit transactions"
            ))
        else:
            account_error = validate_account_format(transaction.toAccount, "toAccount")
            if account_error:
                errors.append(account_error)

    elif transaction.type == TransactionType.WITHDRAWAL:
        # Withdrawal requires fromAccount
        if not transaction.fromAccount:
            errors.append(ValidationErrorDetail(
                field="fromAccount",
                message="fromAccount is required for withdrawal transactions"
            ))
        else:
            account_error = validate_account_format(transaction.fromAccount, "fromAccount")
            if account_error:
                errors.append(account_error)

    elif transaction.type == TransactionType.TRANSFER:
        # Transfer requires both accounts
        if not transaction.fromAccount:
            errors.append(ValidationErrorDetail(
                field="fromAccount",
                message="fromAccount is required for transfer transactions"
            ))
        else:
            account_error = validate_account_format(transaction.fromAccount, "fromAccount")
            if account_error:
                errors.append(account_error)

        if not transaction.toAccount:
            errors.append(ValidationErrorDetail(
                field="toAccount",
                message="toAccount is required for transfer transactions"
            ))
        else:
            account_error = validate_account_format(transaction.toAccount, "toAccount")
            if account_error:
                errors.append(account_error)

        # Check that from and to accounts are different
        if transaction.fromAccount and transaction.toAccount:
            if transaction.fromAccount == transaction.toAccount:
                errors.append(ValidationErrorDetail(
                    field="toAccount",
                    message="fromAccount and toAccount must be different for transfers"
                ))

    return errors
